### PR TITLE
New Shared Hooks Module

### PR DESCRIPTION
**Pull request by jakemhiller**
Includes the two generic hooks we already had in the codebase, I'm sure there are more in there that we could move here.

I plan on borrowing heavily from stuff like [react-use](https://github.com/streamich/react-use) and moving more useful hooks in here as we need them.

### CODE REVIEW

File: packages/hooks/__tests__/useStorageState-test.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+const renderUseStorageHook = () =>
+  renderHook(() => useStorageState(STORAGE_KEY, STORAGE_VALUE, localStorage));
+
+const renderUseSessionStorageHook = () =>
+  renderHook(() => useSessionStorageState(STORAGE_KEY, STORAGE_VALUE));
+
+const renderUseLocalStorageHook = () =>
+  renderHook(() => useLocalStorageState(STORAGE_KEY, STORAGE_VALUE));
+describe('useStorageState', () => {
```
Comment (JoshuaKGoldberg) [inline packages/hooks/__tests__/useStorageState-test.ts]: The teeniest of the nits: blank line? 😄

File: packages/hooks/src/useStorageState.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  defaultValue: T
+): [T, (newValue: T) => void] {
+  if (isServer()) {
+    return [defaultValue, () => {}];
+  }
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  return useStorageState(key, defaultValue, sessionStorage);
+}
+
+export function useLocalStorageState<T>(
```
Comment (JoshuaKGoldberg) [inline packages/hooks/src/useStorageState.ts]: What are your thoughts on marking these all as `@deprecated`, and instead using the ones from `react-use`? I skimmed the list and it looked like they have equivalents.

```suggestion
/**
 * @deprecated Use {XYZ} from `react-use` instead.
 */
export function useLocalStorageState<T>(
```

File: packages/hooks/package.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "@codecademy/hooks",
+  "description": "Library of Various Reusable React Hooks",
+  "version": "1.0.0",
+  "author": "Codecademy Engineering <dev@codecademy.com>",
+  "engines": {
+    "node": ">=6.0.0",
```
Comment (JoshuaKGoldberg) [inline packages/hooks/package.json]: Heh, such back-support...

File: packages/hooks/__tests__/usePrevious-test.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    const { result, rerender } = setUp();
+
+    rerender({ state: 2 });
+    expect(result.current).toBe(0);
+
+    rerender({ state: 4 });
+    expect(result.current).toBe(2);
+
+    rerender({ state: 6 });
+    expect(result.current).toBe(4);
```
Comment (JoshuaKGoldberg) [inline packages/hooks/__tests__/usePrevious-test.ts]: Hmm, this feels weird (non-AAA-y) as a unit tests... I would normally prefer this as 2-3 separate tests for each piece of functionality. Thoughts?

File: packages/hooks/README.md [hunk 5]
```diff
@@ -0,0 +1 @@
+# React Hooks
```
Comment (JoshuaKGoldberg) [inline packages/hooks/README.md]: Up for adding a little description here?

File: packages/hooks/src/useStorageState.ts (hunk 2 above)
Comment (jakemhiller) [reply packages/hooks/src/useStorageState.ts]: why would i do that if we don't use these yet?

Comment (jakemhiller) [reply packages/hooks/src/useStorageState.ts]: seems like i should just close this and we'll use `react-use` in the main app. i can make those changes in my other branch

**Pull request closed without merging**